### PR TITLE
pipe: allow queueing pending handles

### PR DESCRIPTION
**Pull request by indutny**
Call read2_cb multiple times for queued fds.

/cc @saghul @piscisaureus

### CODE REVIEW

**Comment by indutny** [comment]
I'll work on test soon.

**Comment by indutny** [comment]
Nvm, I don't think that it is really possible...

File: src/unix/stream.c [hunk 1]
```diff
@@ -951,6 +973,40 @@ static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
...
+                                       sizeof(*stream->queued_fds));
+      if (stream->queued_fds == NULL)
+        return UV_ENOMEM;
+    }
+  }
+
+  /* Put fd in a queue */
+  stream->queued_fds[2 + queue_offset++] = fd;
+  stream->queued_fds[0] = queue_offset;
+  stream->queued_fds[1] = queue_len;
```
Comment (indutny) [inline src/unix/stream.c:1004]: Hm... this write isn't necessary, I'll move it above.

File: src/unix/stream.c [hunk 2]
```diff
@@ -958,6 +1014,17 @@ static void uv__read(uv_stream_t* stream) {
   struct cmsghdr* cmsg;
   char cmsg_space[64];
   int count;
+  int err;
+
+  /* Has queued fds */
+  if (stream->accepted_fd >= 0) {
+    uv_buf_t buf = { NULL, 0 };
```
Comment (saghul) [inline src/unix/stream.c:1021]: Hum, I thought now we could call write2 and send some data and a handle. Wouldn't you be ignoring the received data here?

**Comment by saghul** [comment]
The approach LGTM. Does Windows support any of this?

File: src/unix/stream.c (hunk 2 above)
Comment (indutny) [inline src/unix/stream.c:1021]: Hm... right I would, perhaps, I just need to remove return.

**Comment by txdv** [comment]
So does it work or not?

**Comment by indutny** [comment]
@saghul I don't think that windows needs it, but it does support receiving handles on IPC pipe.

File: src/unix/stream.c (hunk 2 above)
Comment (saghul) [inline src/unix/stream.c:1021]: Could be yeah.

Comment (saghul) [inline src/unix/stream.c:1021]: Perhaps the buf can be defined as static?

File: src/unix/stream.c [hunk 3]
```diff
@@ -958,6 +1014,17 @@ static void uv__read(uv_stream_t* stream) {
   struct cmsghdr* cmsg;
   char cmsg_space[64];
   int count;
+  int err;
+
+  /* Has queued fds */
+  if (stream->accepted_fd >= 0) {
```
Comment (saghul) [inline src/unix/stream.c:1020]: I like stream->accepted_fd != -1 better :-)

**Comment by saghul** [comment]
> @saghul I don't think that windows needs it, but it does support receiving handles on IPC pipe.

I mean, if there is any way to get more than one at the same time.

**Comment by indutny** [comment]
@saghul I have no idea, but I think no. Otherwise @piscisaureus would had told me about it yesterday.

**Comment by indutny** [comment]
Thank you, landed in 08aeaf6

**Comment by saghul** [comment]
No test?

**Comment by indutny** [comment]
@saghul yep, it's quite hard to get it queuing handles...

**Comment by saghul** [comment]
Yeah. I believe you then :-)

File: src/unix/stream.c [hunk 4]
```diff
@@ -951,6 +973,40 @@ static void uv__stream_eof(uv_stream_t* stream, const uv_buf_t* buf) {
...
+    queue_offset = stream->queued_fds[0];
+    queue_len = stream->queued_fds[1];
+
+    /* Grow */
+    if (queue_offset == queue_len) {
+      queue_len += 8;
+      stream->queued_fds = realloc(stream->queued_fds,
+                                   (queue_len + 2) *
+                                       sizeof(*stream->queued_fds));
+      if (stream->queued_fds == NULL)
```
Comment (bnoordhuis) [inline src/unix/stream.c:996]: Leaks memory.

File: src/unix/stream.c [hunk 5]
```diff
@@ -1493,6 +1563,8 @@ void uv__stream_close(uv_stream_t* handle) {
     handle->accepted_fd = -1;
   }
 
+  free(handle->queued_fds);
```
Comment (bnoordhuis) [inline src/unix/stream.c:1566]: Why don't you NULL it?  Helps catch use-after-free bugs.

File: include/uv-unix.h [hunk 6]
```diff
@@ -229,6 +229,7 @@ struct uv__work {
   uv_connection_cb connection_cb;                                             \
   int delayed_error;                                                          \
   int accepted_fd;                                                            \
+  int* queued_fds;                                                            \
```
Comment (bnoordhuis) [inline include/uv-unix.h:232]: I don't really understand why you didn't use a (possibly private) struct for this, the queue's logic is awfully implicit now.

File: src/unix/stream.c [hunk 7]
```diff
@@ -586,9 +586,31 @@ int uv_accept(uv_stream_t* server, uv_stream_t* client) {
...
+
+    /* All read, free */
+    if (--server->queued_fds[0] == 0) {
+      free(server->queued_fds);
+      server->queued_fds = NULL;
+    } else {
+      /* Shift rest */
+      memmove(server->queued_fds + 2,
+              server->queued_fds + 3,
+              server->queued_fds[0]);
```
Comment (bnoordhuis) [inline src/unix/stream.c:603]: This memmove() doesn't do what you think it does.

Comment (indutny) [inline src/unix/stream.c:603]: Oh gosh, I forgot `sizeof()`

**Comment by bnoordhuis** [comment]
I gather that the purpose of this change is to handle recvmsg()s that return multiple file descriptors?  How about changing the cmsg_space buffer so it only has room for a single file descriptor?

**Comment by indutny** [comment]
Can't say I wasn't thinking about it, but what if some other message will sneak in?

**Comment by indutny** [comment]
Temporarly reverted.

**Comment by indutny** [comment]
@bnoordhuis @saghul pushed fixes. let's review and land it!

**Comment by indutny** [comment]
@bnoordhuis additional comment about cmsg_space, what if not a libuv is writing to the other side of the pipe? I think we would support it this way, but definitely won't your way.

**Comment by bnoordhuis** [comment]
You mean what if some non-libuv process is sending multiple file descriptors in a message but libuv's cmsg receive buffer is only large enough for one?  It depends on the platform:
- Linux returns the first file descriptor and closes the others.
- OS X returns the first file descriptor and closes the others.  cmsg->cmsg_len is the number of bytes that the kernel would have needed to return all file descriptors, not the actual number of bytes written (in other words, you have to be careful not to overrun the receive buffer.)
- OpenBSD fails with EMSGSIZE and closes all file descriptors.

I don't know what FreeBSD, NetBSD and Solaris do.  I guess the point is that you can't win: if the receive buffer is too small, you lose some or all file descriptors.  If you make it too big, you become vulnerable to (possibly unwitting) EMFILE denial-of-service attacks.

**Comment by indutny** [comment]
Well, it couldn't be a true denial-of-service since it's inter-process communication anyway, and we're in trustful environment. I don't think that we should make the buffer significantly larger or even making it bigger at all. I just think that we should handle such cases and be able to queue file descriptors for a bit. This makes clustering in C on solaris much more simpler and faster.

File: include/uv-unix.h (hunk 6 above)
Comment (jclulow) [inline include/uv-unix.h:232]: Yeah, I totally agree with @bnoordhuis -- this should be a `struct`.  Doing it with the int array requires surprising code, and is not self-describing or easily debuggable.

Comment (indutny) [inline include/uv-unix.h:232]: @jclulow please take a look at https://github.com/joyent/libuv/pull/1056/files

**Comment by bnoordhuis** [comment]
> I just think that we should handle such cases and be able to queue file descriptors for a bit.

Yeah, but it will only ever a so-so solution because it breaks down the moment you receive more file descriptors than anticipated (in platform-specific ways no less, as noted above.)

There's also the issue that your PR doesn't quite work as advertised because libuv still only pulls the first file descriptor from the SCM_RIGHTS message, the others are leaked.  (Long-standing issue that I never got around to fixing.)

**Comment by indutny** [comment]
@bnoordhuis may I ask you to continue on https://github.com/joyent/libuv/pull/1056/files ? Also, I don't see why it leaks excessive fds right now... may you shed some light on it, please?

**Comment by bnoordhuis** [comment]
> Also, I don't see why it leaks excessive fds right now... may you shed some light on it, please?

The struct msghdr that's passed to recvmsg() has msg_controllen set to 64.  That's big enough to receive 11 or 12 file descriptors but libuv only reads the first one.

Put another way, it scans for the SCM_RIGHTS message but when it finds it, it only considers the first data element.  It needs to check cmsg->cmsg_len.

**Pull request closed without merging**